### PR TITLE
[4701] Add ability to remove access to a provider or lead school

### PR DESCRIPTION
**Pull request by jhenderson**
### Context
https://trello.com/c/xKsNkiQE/4701-m-support-add-ability-to-remove-access-to-a-provider

### Changes proposed in this pull request
- Add two new tables in admin user show page for providers and lead schools
- Each row in table has a link to remove access
- Clicking on remove access link brings user to confirmation page

### Guidance to review
- Sign in a system admin
- Visit a user show page
- Add a provider
- Click "Remove access" link and follow through
- Provider should be remove from the list
- Repeat for lead schools

### Screenshot
Screenshot 2022-09-23 at 15 58 57

### Important business

- [ ] Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?
- [ ] Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?

NB: Please notify the #twd_data_insights team and ask for a review if new fields are being added to analytics.yml

### CODE REVIEW

**Comment by jhenderson** [comment]
[[M] Support: add ability to remove access to a provider](https://trello.com/c/xKsNkiQE/4701-m-support-add-ability-to-remove-access-to-a-provider)

**Review by markhallen — APPROVED** [review]
You've Got LGTM

**Comment by claire-hughez** [comment]
This looks good to me!

File: app/views/system_admin/lead_schools/accessions/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds-from-desktop">
+    <span class="govuk-caption-l"><%= @user.name %></span>
+    <h1 class="govuk-heading-l">Remove <%= @user.name %>’s access to <%= @lead_school.name %></h1>
+    <%= govuk_button_to "Yes I’m sure – remove #{@user.name}’s access to #{@lead_school.name}",
+                        user_lead_school_accessions_path(@user, @lead_school),
+                        method: :delete,
+                        class: "govuk-button govuk-button--warning" %>
+    <%= govuk_link_to("Cancel", user_path(@user)) %>
```
Comment (edwardhorsford) [inline app/views/system_admin/lead_schools/accessions/edit.html.erb:11]:
```suggestion
    <%= govuk_link_to("Cancel", user_path(@user), class: "govuk-link--no-visited-state govuk-body") %>
```

File: app/views/system_admin/users/show.html.erb [hunk 2]
```diff
@@ -1,69 +1,84 @@
...
+            <td class="govuk-table__cell">
+              <%= govuk_link_to provider.name, provider_path(provider), class: 'provider' %>
+            </td>
+            <td class="govuk-table__cell">
+              <%= govuk_link_to "Remove", edit_user_provider_accessions_path(@user, provider), class: 'remove-access' %>
+            </td>
+          </tr>
+        <% end %>
+      </tbody>
+    </table>
```
Comment (edwardhorsford) [inline app/views/system_admin/users/show.html.erb:46]: Can we pull out `Providers` as a heading, and only show the table where there are providers? When there are no providers show:

```
<p class="govuk-body govuk-hint">User is not a member of any providers</p>
```


Example from prototype:
Screenshot 2022-09-28 at 15 15 44

File: app/views/system_admin/users/show.html.erb [hunk 3]
```diff
@@ -1,69 +1,84 @@
...
+            <td class="govuk-table__cell">
+              <%= govuk_link_to lead_school.name, lead_school_path(lead_school), class: "lead-school" %>
+            </td>
+            <td class="govuk-table__cell">
+              <%= govuk_link_to "Remove", edit_user_lead_school_accessions_path(@user, lead_school), class: 'remove-access' %>
+            </td>
+          </tr>
+        <% end %>
+      </tbody>
+    </table>
```
Comment (edwardhorsford) [inline app/views/system_admin/users/show.html.erb:75]: Can we pull out `Lead schools` as a heading, and only show the table where there are lead schools? When there are no lead schools show:

```
<p class="govuk-body govuk-hint">User is not a member of any lead schools</p>
```


Example from prototype:
Screenshot 2022-09-28 at 15 15 44

File: app/views/system_admin/users/show.html.erb [hunk 4]
```diff
@@ -1,69 +1,84 @@
...
+    <%= render GovukButtonLinkTo::View.new(body: "Add a user to a lead school",
```
Comment (edwardhorsford) [inline app/views/system_admin/users/show.html.erb:76]:
```suggestion
    <%= render GovukButtonLinkTo::View.new(body: "Add user to a lead school",
```

File: app/views/system_admin/users/show.html.erb [hunk 5]
```diff
@@ -1,69 +1,84 @@
...
+              <%= govuk_link_to provider.name, provider_path(provider), class: 'provider' %>
+            </td>
+            <td class="govuk-table__cell">
+              <%= govuk_link_to "Remove", edit_user_provider_accessions_path(@user, provider), class: 'remove-access' %>
+            </td>
+          </tr>
+        <% end %>
+      </tbody>
+    </table>
+    <%= render GovukButtonLinkTo::View.new(body: "Add a user to a provider",
```
Comment (edwardhorsford) [inline app/views/system_admin/users/show.html.erb:47]:
```suggestion
    <%= render GovukButtonLinkTo::View.new(body: "Add user to a provider",
```

Sorry, this was poor copy in the prototype.

File: app/views/system_admin/users/show.html.erb [hunk 6]
```diff
@@ -1,69 +1,84 @@
...
+      </tr>
+      </thead>
+      <tbody class="govuk-table__body">
+        <% @user.providers.each do |provider| %>
+          <tr class="govuk-table__row">
+            <td class="govuk-table__cell">
+              <%= govuk_link_to provider.name, provider_path(provider), class: 'provider' %>
+            </td>
+            <td class="govuk-table__cell">
+              <%= govuk_link_to "Remove", edit_user_provider_accessions_path(@user, provider), class: 'remove-access' %>
```
Comment (edwardhorsford) [inline app/views/system_admin/users/show.html.erb:41]: This link will need hidden text to identify it uniquely when there are multiple remove links. I think we might already have a method for doing hidden text?

Ideally something like:

`Remove<span class="govuk-visually-hidden"> access to {{provider.name}}</span>`

File: app/views/system_admin/users/show.html.erb [hunk 7]
```diff
@@ -1,69 +1,84 @@
...
+      </tr>
+      </thead>
+      <tbody class="govuk-table__body">
+        <% @user.lead_schools.each do |lead_school| %>
+          <tr class="govuk-table__row">
+            <td class="govuk-table__cell">
+              <%= govuk_link_to lead_school.name, lead_school_path(lead_school), class: "lead-school" %>
+            </td>
+            <td class="govuk-table__cell">
+              <%= govuk_link_to "Remove", edit_user_lead_school_accessions_path(@user, lead_school), class: 'remove-access' %>
```
Comment (edwardhorsford) [inline app/views/system_admin/users/show.html.erb:70]: This link will need hidden text to identify it uniquely when there are multiple remove links. I think we might already have a method for doing hidden text?

Ideally something like:

`Remove<span class="govuk-visually-hidden"> access to {{lead_school.name}}</span>`

File: app/views/system_admin/users/show.html.erb [hunk 8]
```diff
@@ -1,69 +1,84 @@
...
+    <%= render GovukButtonLinkTo::View.new(body: "Add a user to a provider",
+                                           url: new_user_provider_path(@user),
+                                           class_option: "govuk-button--secondary add-user-to-provider") %>
+  </div>
+</div>
+
+<div class="govuk-grid-row">
+  <div class ="govuk-grid-column-full">
+    <table id="lead-schools" class="govuk-table">
+      <caption class="govuk-table__caption govuk-table__caption--m">Lead Schools</caption>
```
Comment (edwardhorsford) [inline app/views/system_admin/users/show.html.erb:56]:
```suggestion
      <caption class="govuk-table__caption govuk-table__caption--m">Lead schools</caption>
```

File: app/views/system_admin/lead_schools/accessions/edit.html.erb [hunk 9]
```diff
@@ -0,0 +1,13 @@
+<%= render PageTitle::View.new(text: "Yes I’m sure – remove #{@user.name}’s access to #{@lead_school.name}") %>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds-from-desktop">
+    <span class="govuk-caption-l"><%= @user.name %></span>
+    <h1 class="govuk-heading-l">Remove <%= @user.name %>’s access to <%= @lead_school.name %></h1>
```
Comment (edwardhorsford) [inline app/views/system_admin/lead_schools/accessions/edit.html.erb:6]:
```suggestion
    <h1 class="govuk-heading-l">Remove <%= @user.name %>’s access to <%= @lead_school.name %>?</h1>
```

File: app/views/system_admin/providers/accessions/edit.html.erb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+<%= render PageTitle::View.new(text: "Yes I’m sure – remove #{@user.name}’s access to #{@provider.name}") %>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds-from-desktop">
+    <span class="govuk-caption-l"><%= @user.name %></span>
+    <h1 class="govuk-heading-l">Remove <%= @user.name %>’s access to <%= @provider.name %></h1>
```
Comment (edwardhorsford) [inline app/views/system_admin/providers/accessions/edit.html.erb:6]:
```suggestion
    <h1 class="govuk-heading-l">Remove <%= @user.name %>’s access to <%= @provider.name %>?</h1>
```

**Review by edwardhorsford — CHANGES_REQUESTED** [review]
Looking good! a few more bits to do

**Comment by jhenderson** [comment]
@edwardhorsford I've updated the PR with all of your suggestions. Thanks.

File: app/views/system_admin/providers/accessions/edit.html.erb [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds-from-desktop">
+    <span class="govuk-caption-l"><%= @user.name %></span>
+    <h1 class="govuk-heading-l">Remove <%= @user.name %>'s access to <%= @provider.name %>?</h1>
+    <%= govuk_button_to "Yes I'm sure – remove #{@user.name}'s access to #{@provider.name}",
+                        user_provider_accessions_path(@user, @provider),
+                        method: :delete,
+                        class: "govuk-button govuk-button--warning" %>
+    <%= govuk_link_to("Cancel", user_path(@user)) %>
```
Comment (edwardhorsford) [inline app/views/system_admin/providers/accessions/edit.html.erb:11]:
```suggestion
    <%= govuk_link_to("Cancel", user_path(@user), class: "govuk-link--no-visited-state govuk-body") %>
```

File: app/views/system_admin/lead_schools/accessions/edit.html.erb [hunk 12]
```diff
@@ -0,0 +1,13 @@
+<%= render PageTitle::View.new(text: "Yes I'm sure – remove #{@user.name}'s access to #{@lead_school.name}") %>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds-from-desktop">
+    <span class="govuk-caption-l"><%= @user.name %></span>
+    <h1 class="govuk-heading-l">Remove <%= @user.name %>'s access to <%= @lead_school.name %>?</h1>
+    <%= govuk_button_to "Yes I'm sure – remove #{@user.name}'s access to #{@lead_school.name}",
```
Comment (edwardhorsford) [inline app/views/system_admin/lead_schools/accessions/edit.html.erb:7]:
```suggestion
    <%= govuk_button_to "Yes I’m sure – remove #{@user.name}’s access to #{@lead_school.name}",
```

File: app/views/system_admin/lead_schools/accessions/edit.html.erb [hunk 13]
```diff
@@ -0,0 +1,13 @@
+<%= render PageTitle::View.new(text: "Yes I'm sure – remove #{@user.name}'s access to #{@lead_school.name}") %>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds-from-desktop">
+    <span class="govuk-caption-l"><%= @user.name %></span>
+    <h1 class="govuk-heading-l">Remove <%= @user.name %>'s access to <%= @lead_school.name %>?</h1>
```
Comment (edwardhorsford) [inline app/views/system_admin/lead_schools/accessions/edit.html.erb:6]:
```suggestion
    <h1 class="govuk-heading-l">Remove <%= @user.name %>’s access to <%= @lead_school.name %>?</h1>
```

File: app/views/system_admin/lead_schools/accessions/edit.html.erb [hunk 14]
```diff
@@ -0,0 +1,13 @@
+<%= render PageTitle::View.new(text: "Yes I'm sure – remove #{@user.name}'s access to #{@lead_school.name}") %>
```
Comment (edwardhorsford) [inline app/views/system_admin/lead_schools/accessions/edit.html.erb:1]:
```suggestion
<%= render PageTitle::View.new(text: "Yes I’m sure – remove #{@user.name}’s access to #{@lead_school.name}") %>
```

File: app/views/system_admin/providers/accessions/edit.html.erb [hunk 15]
```diff
@@ -0,0 +1,13 @@
+<%= render PageTitle::View.new(text: "Yes I'm sure – remove #{@user.name}'s access to #{@provider.name}") %>
```
Comment (edwardhorsford) [inline app/views/system_admin/providers/accessions/edit.html.erb:1]:
```suggestion
<%= render PageTitle::View.new(text: "Yes I’m sure – remove #{@user.name}’s access to #{@provider.name}") %>
```

File: app/views/system_admin/providers/accessions/edit.html.erb [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds-from-desktop">
+    <span class="govuk-caption-l"><%= @user.name %></span>
+    <h1 class="govuk-heading-l">Remove <%= @user.name %>'s access to <%= @provider.name %>?</h1>
```
Comment (edwardhorsford) [inline app/views/system_admin/providers/accessions/edit.html.erb:6]:
```suggestion
    <h1 class="govuk-heading-l">Remove <%= @user.name %>’s access to <%= @provider.name %>?</h1>
```

File: app/views/system_admin/providers/accessions/edit.html.erb [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+    <%= govuk_button_to "Yes I'm sure – remove #{@user.name}'s access to #{@provider.name}",
```
Comment (edwardhorsford) [inline app/views/system_admin/providers/accessions/edit.html.erb:7]:
```suggestion
    <%= govuk_button_to "Yes I’m sure – remove #{@user.name}’s access to #{@provider.name}",
```

File: app/views/system_admin/users/show.html.erb [hunk 18]
```diff
@@ -1,69 +1,100 @@
...
+<%= render GovukButtonLinkTo::View.new(body: "Add user to a provider",
+                                       url: new_user_provider_path(@user),
+                                       class_option: "govuk-button--secondary add-user-to-provider") %>
+
+<h2 class="govuk-heading-m">Lead schools</h2>
+<% if @user.lead_schools.any? %>
+  <div class="govuk-grid-row">
+    <div class ="govuk-grid-column-full">
+      <table id="lead-schools" class="govuk-table">
+        <caption class="govuk-visually-hidden govuk-table__caption govuk-table__caption--m">Lead schools</caption>
```
Comment (edwardhorsford) [inline app/views/system_admin/users/show.html.erb:65]:
```suggestion
```

File: app/views/system_admin/users/show.html.erb [hunk 19]
```diff
@@ -1,69 +1,100 @@
...
-      <% if @user.providers.any? %>
-        <div class="govuk-!-margin-bottom-3">
+</dl>
+
+<h2 class="govuk-heading-m">Providers</h2>
+<% if @user.providers.any? %>
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-full">
+      <table id="providers" class="govuk-table">
+        <caption class="govuk-visually-hidden govuk-table__caption govuk-table__caption--m">Providers</caption>
```
Comment (edwardhorsford) [inline app/views/system_admin/users/show.html.erb:29]:
```suggestion
```

I don't think these captions are needed as we have headings that describe the content.

File: app/views/system_admin/users/show.html.erb [hunk 20]
```diff
@@ -1,69 +1,100 @@
...
+        <thead class="govuk-table__head">
+          <tr class="govuk-table__row">
+            <th scope="col" class="govuk-table__header">Name</th>
+            <th scope="col" class="govuk-table__header">Remove access</th>
```
Comment (edwardhorsford) [inline app/views/system_admin/users/show.html.erb:33]:
```suggestion
            <th scope="col" class="govuk-table__header govuk-!-width-one-third">Remove access</th>
```

File: app/views/system_admin/users/show.html.erb [hunk 21]
```diff
@@ -1,69 +1,100 @@
...
+<h2 class="govuk-heading-m">Lead schools</h2>
+<% if @user.lead_schools.any? %>
+  <div class="govuk-grid-row">
+    <div class ="govuk-grid-column-full">
+      <table id="lead-schools" class="govuk-table">
+        <caption class="govuk-visually-hidden govuk-table__caption govuk-table__caption--m">Lead schools</caption>
+        <thead class="govuk-table__head">
+          <tr class="govuk-table__row">
+            <th scope="col" class="govuk-table__header">Name</th>
+            <th scope="col" class="govuk-table__header">Remove access</th>
```
Comment (edwardhorsford) [inline app/views/system_admin/users/show.html.erb:69]:
```suggestion
            <th scope="col" class="govuk-table__header govuk-!-width-one-third">Remove access</th>
```

**Review by edwardhorsford — CHANGES_REQUESTED** [review]

File: app/views/system_admin/users/show.html.erb (hunk 19 above)
Comment (jhenderson) [reply app/views/system_admin/users/show.html.erb:29]: I did try to remove it, but SonarCloud complained that a table should have a caption. It seems all our tables have a caption.

**Review by edwardhorsford — APPROVED** [review]

**Pull request merged**